x/dymns/keeper/dym_name.go

```diff
@@ -0,0 +1,800 @@
+// Returns nil if Dym-Name does not exist or is expired.
+func (k Keeper) GetDymNameWithExpirationCheck(ctx sdk.Context, name string) *dymnstypes.DymName {
+	// Legacy TODO DymNS: always use this on queries
+	dymName := k.GetDymName(ctx, name)
+	if dymName == nil {
+		return nil
+	}
+
+	if dymName.IsExpiredAtCtx(ctx) {
+		return nil
+	}
+
+	return dymName
+}
+
+// DeleteDymName removes a Dym-Name from the KVStore.
+// This function will remove the Dym-Name record as well as the reverse mappings records.
+func (k Keeper) DeleteDymName(ctx sdk.Context, name string) error {
+	if err := k.BeforeDymNameOwnerChanged(ctx, name); err != nil {
+		return err
+	}
+
+	if err := k.BeforeDymNameConfigChanged(ctx, name); err != nil {
+		return err
+	}
+
+	store := ctx.KVStore(k.storeKey)
+	dymNameKey := dymnstypes.DymNameKey(name)
+	store.Delete(dymNameKey)
+
+	return nil
+}
+
+// GetAllNonExpiredDymNames returns all non-expired Dym-Names from the KVStore.
+// This function will filter out all expired Dym-Names based on the time of the context.
+func (k Keeper) GetAllNonExpiredDymNames(ctx sdk.Context) (list []dymnstypes.DymName) {
+	store := ctx.KVStore(k.storeKey)
+
+	iterator := sdk.KVStorePrefixIterator(store, dymnstypes.KeyPrefixDymName)
+	defer func() {
+		_ = iterator.Close()
+	}()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var dymName dymnstypes.DymName
+		k.cdc.MustUnmarshal(iterator.Value(), &dymName)
+
+		if dymName.IsExpiredAtCtx(ctx) {
+			continue
+		}
+
+		list = append(list, dymName)
+	}
+
+	return list
+}
+
+// GetAllDymNames returns all Dym-Names from the KVStore.
+// No filter applied, to be used in genesis export.
+func (k Keeper) GetAllDymNames(ctx sdk.Context) (list []dymnstypes.DymName) {
+	store := ctx.KVStore(k.storeKey)
+
+	iterator := sdk.KVStorePrefixIterator(store, dymnstypes.KeyPrefixDymName)
+	defer func() {
+		_ = iterator.Close()
+	}()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var dymName dymnstypes.DymName
+		k.cdc.MustUnmarshal(iterator.Value(), &dymName)
+		list = append(list, dymName)
+	}
+
+	return list
+}
+
+// PruneDymName removes a Dym-Name from the KVStore, as well as all related records.
+func (k Keeper) PruneDymName(ctx sdk.Context, name string) error {
+	// remove SO (force, ignore active SO)
+	k.DeleteSellOrder(ctx, name, dymnstypes.NameOrder)
+
+	// remove historical SO
+	k.DeleteHistoricalSellOrders(ctx, name, dymnstypes.NameOrder)
+	k.SetMinExpiryHistoricalSellOrder(ctx, name, dymnstypes.NameOrder, 0)
+
+	dymName := k.GetDymName(ctx, name)
+	if dymName == nil {
+		return nil
+	}
+
+	// remove config
+	// This seems not necessary because we are going to remove the record anyway,
+	// but just let it here to clear the business logic
+	dymName.Configs = nil   // all configuration should be removed
+	dymName.Owner = ""      // no one owns it anyone
+	dymName.Controller = "" // no one controls it anyone
+
+	return k.DeleteDymName(ctx, name)
```

Comment:
`DeleteDymName` receives string params, because I use the DymName record to emit an event.
So what I want is: when delete it from store, it fires the correct up-to-date data before delete.